src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java

```diff
@@ -976,6 +976,19 @@ public void checkIfProjectExists(ProgrammingExercise programmingExercise) {
         // means the project does not exist in version control server and does not exist in continuous integration server
     }
 
+    /**
+     * Recreates the template and solution build plans by deleting the old ones and creating them again as for a new exercise.
+     *
+     * The student build plans are also cleaned up so that no inconsistencies between the template/solution and the student build plans occur.
+     * The student build plans are not recreated, as that will happen automatically when the student submits the next time or the student resumes the exercise.
+     *
+     * @param programmingExercise The exercise for which the template and solution should be recreated.
+     */
+    public void recreateBuildPlans(final ProgrammingExercise programmingExercise) {
```

Comment:
I guess we should communicate this change to the instructor. In particular that student build plans are deleted and not recreated. The button is just called "Recreate Build Plans" and would indicate something else.
I guess now that all student build plans are deleted, the instructor should also confirm this operation to avoid that he/she clicks the button accidentally. 

In the confirm dialog, we should offer options whether the existing student build plans should be deleted and - if chosen - whether they should be automatically re-created.

While the operation is not really destructive (build plan is automatically recreated and history of results should be stored in Artemis), the history of the student build plan is still deleted. Also, some students might be working on the exercise at this point of time and could have a bad experience if the build plan is suddenly deleted. In large courses, this operation can take quite some time, if deletion is done sequentially and it might also take quite some resources (if performance is a concern here, we could potentially allow to delete multiple build plans at the same time, e.g. by using parallel streams, we already do this in other places)

I guess informing the instructor about these consequences helps for a better decision.

Reply:
I updated the Recreate Build Plans button to now explain in more detail what the operation does and also explicitly ask for confirmation.
The student build plans are no longer deleted by default. Instead, the explanation makes recommendations in which cases the instructor should separately run the Cleanup action as well.
I added screenshots the PR description showing those new messages.